File: lib/barista/translator.rb
# frozen_string_literal: true

module Barista
  # Translates an Intermediate into a valid Caffeine v6 expectations string.
  module Translator
    def self.translate(intermediate)
      return empty_file(intermediate) if intermediate.guarantees.empty?

      lines = ["Expectations"]

      intermediate.guarantees.each do |guarantee|
        lines << ""
        lines << "\"#{guarantee.name}\":"
        lines << "  Guarantees #{format_threshold(guarantee.threshold)}% over #{guarantee.window_days}d window"
      end

      lines << ""
      lines.join("\n")
    end

    def self.format_threshold(threshold)
      int = threshold.to_i
      threshold == int ? int : threshold
    end
    private_class_method :format_threshold

    def self.empty_file(intermediate)
      "# No guarantees found for #{intermediate.provider_name}/#{intermediate.service_name}\n" \
        "# Source: #{intermediate.source_url}\n"
    end
    private_class_method :empty_file
  end
end

File: internal/pipeline/translator.go
package pipeline

import (
	"fmt"
	"strings"
)

func Translate(i *Intermediate) string {
	if len(i.Guarantees) == 0 {
		return fmt.Sprintf("# No guarantees found for %s/%s\n# Source: %s", i.ProviderName, i.ServiceName, i.SourceURL)
	}

	var sb strings.Builder

	sb.WriteString("# === Expectations ===\n")
	fmt.Fprintf(&sb, "Unmeasured Expectations\n")
	fmt.Fprintf(&sb, "  # Source: %s\n", i.SourceURL)

	for _, g := range i.Guarantees {
		sb.WriteString("\n")
		fmt.Fprintf(&sb, "  * %q:\n", toTitleCase(g.Name))
		sb.WriteString("    Provides {\n")
		fmt.Fprintf(&sb, "      threshold: %s%%,\n", formatThreshold(g.Threshold))
		fmt.Fprintf(&sb, "      window_in_days: %d\n", g.WindowDays)
		sb.WriteString("    }")
	}

	return sb.String()
}

func toTitleCase(s string) string {
	words := strings.Split(s, "_")
	for i, w := range words {
		if len(w) > 0 {
			words[i] = strings.ToUpper(w[:1]) + w[1:]
		}
	}
	return strings.Join(words, " ")
}

func formatThreshold(t float64) string {
	if t == float64(int(t)) {
		return fmt.Sprintf("%d", int(t))
	}
	return fmt.Sprintf("%g", t)
}
